velox/substrait/VeloxToSubstraitPlan.cpp

```diff
@@ -170,13 +176,119 @@ void VeloxToSubstraitPlanConvertor::toSubstrait(
+  int64_t groupingKeySize = groupingKeys.size();
+  ::substrait::AggregateRel_Grouping* aggGroupings =
+      aggregateRel->add_groupings();
+
+  for (int64_t i = 0; i < groupingKeySize; i++) {
+    aggGroupings->add_grouping_expressions()->MergeFrom(
+        exprConvertor_->toSubstraitExpr(
+            arena,
+            std::dynamic_pointer_cast<const core::ITypedExpr>(
+                groupingKeys.at(i)),
+            inputType));
+  }
+
+  // AggregatesSize should be equal to or greater than the aggregateMasks Size.
+  // Two cases: 1. aggregateMasksSize = 0, aggregatesSize > aggregateMasksSize.
+  // 2. aggregateMasksSize != 0, aggregatesSize = aggregateMasksSize.
+  auto aggregates = aggregateNode->aggregates();
+  auto aggregateMasks = aggregateNode->aggregateMasks();
+  int64_t aggregatesSize = aggregates.size();
+  int64_t aggregateMasksSize = aggregateMasks.size();
+  VELOX_CHECK_GE(aggregatesSize, aggregateMasksSize);
+
+  for (int64_t i = 0; i < aggregatesSize; i++) {
+    ::substrait::AggregateRel_Measure* aggMeasures =
+        aggregateRel->add_measures();
+
+    auto aggMaskExpr = aggregateMasks.at(i);
+    // Set substrait filter.
+    ::substrait::Expression* aggFilter = aggMeasures->mutable_filter();
+    if (aggMaskExpr.get()) {
+      aggFilter->MergeFrom(exprConvertor_->toSubstraitExpr(
+          arena,
+          std::dynamic_pointer_cast<const core::ITypedExpr>(aggMaskExpr),
+          inputType));
+    } else {
+      // Set null.
+      aggFilter = NULL;
+    }
+
+    // Process measure, eg:sum(a).
+    const auto& aggregatesExpr = aggregates.at(i);
+    ::substrait::AggregateFunction* aggFunction =
+        aggMeasures->mutable_measure();
+
+    // Aggregation function name.
+    const auto& funName = aggregatesExpr->name();
+    // set aggFunction args.
+    const auto& aggregatesExprInputs = aggregatesExpr->inputs();
+
+    int64_t aggregatesExprInputsSize = aggregatesExprInputs.size();
+    for (int64_t j = 0; j < aggregatesExprInputsSize; j++) {
+      ::substrait::Expression* aggFunctionExpr = aggFunction->add_args();
+      core::TypedExprPtr expr = aggregatesExprInputs[j];
+
+      // If the expr is CallTypedExpr, people need do project firstly.
+      if (auto aggregatesExprInput =
+              std::dynamic_pointer_cast<const core::CallTypedExpr>(expr)) {
+        VELOX_NYI("In Velox Plan, the aggregates type cannot be CallTypedExpr");
+      } else if (
+          auto aggregatesExprInput =
+              std::dynamic_pointer_cast<const core::ConstantTypedExpr>(expr) &&
+              funName == "count") {
+        // For case count(1).
+        aggFunction->clear_args();
+      } else {
+        // FieldAccessTypedExpr and Cast.
+        aggFunctionExpr->MergeFrom(
+            exprConvertor_->toSubstraitExpr(arena, expr, inputType));
+      }
+    }
+
+    // Set substrait aggregate Function reference and output type.
+    aggFunction->set_function_reference(functionMap_[funName]);
```

Comment:
Shouldn't we check if funName exists and throw if it doesn't?

Reply:
fix.